shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1514,34 +1587,58 @@ module MakeImpl<InputSig Lang> {
           PrevStage::callEdgeReturn(call, c, ret, _, out, allowsFieldFlow, apa)
         }
 
+        pragma[nomagic]
+        private predicate fwdFlowIntoRet(
+          RetNodeEx ret, FlowState state, CcNoCall cc, ParamNodeOption summaryCtx, TypOption argT,
+          ApOption argAp, Typ t, Ap ap, ApApprox apa
+        ) {
+          fwdFlow(ret, state, cc, summaryCtx, argT, argAp, t, ap, apa)
+        }
+
         pragma[nomagic]
         private predicate fwdFlowOutCand(
-          DataFlowCall call, DataFlowCallable inner, NodeEx out, FlowState state,
-          ParamNodeOption summaryCtx, TypOption argT, ApOption argAp, Typ t, Ap ap, ApApprox apa
+          DataFlowCall call, RetNodeEx ret, CcNoCall innercc, DataFlowCallable inner, NodeEx out,
+          ApApprox apa, boolean allowsFieldFlow
         ) {
-          exists(RetNodeEx ret, boolean allowsFieldFlow, CcNoCall innercc |
-            fwdFlow(ret, state, innercc, summaryCtx, argT, argAp, t, ap, apa) and
-            inner = ret.getEnclosingCallable() and
-            (
-              inner = viableImplCallContextReducedReverseInlineLate(call, innercc) and
-              flowOutOfCallApaInlineLate(call, inner, ret, out, allowsFieldFlow, apa)
-              or
-              flowOutOfCallApaNotCallContextReduced(call, inner, ret, out, allowsFieldFlow, apa,
-                innercc)
-            )
-          |
-            if allowsFieldFlow = false then ap instanceof ApNil else any()
+          fwdFlowIntoRet(ret, _, innercc, _, _, _, _, _, apa) and
+          inner = ret.getEnclosingCallable() and
+          (
+            inner = viableImplCallContextReducedReverseInlineLate(call, innercc) and
+            flowOutOfCallApaInlineLate(call, inner, ret, out, allowsFieldFlow, apa)
+            or
+            flowOutOfCallApaNotCallContextReduced(call, inner, ret, out, allowsFieldFlow, apa,
+              innercc)
           )
         }
 
         pragma[nomagic]
+        private predicate fwdFlowOutValidEdge(
+          DataFlowCall call, RetNodeEx ret, CcNoCall innercc, DataFlowCallable inner, NodeEx out,
+          CcNoCall outercc, ApApprox apa, boolean allowsFieldFlow
+        ) {
+          fwdFlowOutCand(call, ret, innercc, inner, out, apa, allowsFieldFlow) and
+          FwdTypeFlow::typeFlowValidEdgeOut(call, inner) and
```

Comment:
Right, so just double-checking: You're choosing to make this non-linear join smaller by re-joining with `fwdFlowIntoRet` in a subsequent and additional non-linear join? It's a bit counter-intuitive at first, but presumably that's worth it?

Reply:
Indeed.